### DISCUSSION TITLE
Recommendation to protect certain docs with auth (i.e. restrict content)

### DISCUSSION THREAD

**Discussion by nktnet1** [fuma-nama/fumadocs#777] [Q&A] [answered]
Is there a recommendation way for protecting some documentation using custom logic, and perhaps tags? (i.e. display only the relevant content on the side bar)

For example, the user must be logged in or has the correct role (authentication/authorisation).

Ideally have the contents of the protected page not searchable either.

**Comment by fuma-nama** [comment]
you can use Next.js middleware, and manipulate the page tree passed to your docs layout to hide certain pages. Same for search API

**Reply by nktnet1** [reply]
Thanks for replying @fuma-nama!

I was able to add the middleware and tweak the `Page` function to hide the contents, e.g. for a default project made with
 `npm create fumadocs-app`, in `/app/docs/[[...slug]]/page.tsx`:
```tsx
export default async function Page({ params }: { params: { slug?: string[] } }) {
  const page = getPage(params.slug);
  
  // Async function from next-auth
  const session = await getServerSession();

  if (page == null) {
    return notFound();
  }

  // Restrict the content here
  if (!session && !publicContentPaths.includes(page.url)) {
    return (
      <Forbidden />
    );
  }

  const MDX = page.data.exports.default;

  return (
    <DocsPage
      tableOfContent={{ single: true, style: 'clerk' }}
      toc={page.data.exports.toc}
      full={page.data.full}
    >
      <DocsTitle>{page.data.title}</DocsTitle>
      <DocsDescription>{page.data.description}</DocsDescription>
      <DocsBody>
        <MDX />
      </DocsBody>
    </DocsPage>
  );
}
```

However, I'm not too where to tweak the Page Tree - would this be in `/app/docs/source.ts`?

With the search API, since [getServerSession (from next-auth)](https://next-auth.js.org/configuration/nextjs#getserversession) is an `async` function with a request scope, I couldn't find a way to integrate them together. I'm quite new to NextJS and fumadocs so apologies if these things are obvious!

**Reply by fuma-nama** [reply]
on your docs layout, you change the passed page tree.

for Search API, just refer to the `route.ts`, make another GET handler: `export function GET(req: NextRequest)` that wraps the original search API handler. Then you can add your own logic over there

**Reply by nktnet1** [reply]
Thanks @fuma-nama - I was able to get the search feature working, e.g. something like

**Edit: move code to:**
- https://github.com/fuma-nama/fumadocs/discussions/818#discussioncomment-10567821

I did run into an issue with the search feature not working when we specify a `basePath` in our NestJS config though (this was before I made the above changes) - do you know of any way around this? Getting 404, very similar to this [next-auth issue](https://github.com/nextauthjs/next-auth/issues/689#issuecomment-696146462).

![image](https://github.com/user-attachments/assets/1631c0d6-9e2b-4df9-9bea-36f6131b9b9f)


If it's a bug, I'm happy to also file a bug report.

---

> on your docs layout, you change the passed page tree.

Sorry - could you please elaborate on this :(? Are the changes to be made in this file?

`/app/docs/layout.tsx`:
```tsx
import { DocsLayout } from 'fumadocs-ui/layout';
import type { ReactNode } from 'react';
import { docsOptions } from '../layout.config';

export default function Layout({ children }: { children: ReactNode }) {
  return (
    <DocsLayout
      {...docsOptions}
      links={[
        {
          url: '/',
          text: 'Go Home',
        },
      ]}
    >
      {children}
    </DocsLayout>
  );
}
```

I've looked at https://fumadocs.vercel.app/docs/headless/page-tree but couldn't find any way to get the existing tree for filtering. Unless you're suggesting to manually write relevant element?

**Reply by fuma-nama** [reply, marked as answer]
See `docsOptions`, you can change the passed page tree.

And for the base path, you have to tell the search client to pickup the new API endpoint.

```tsx
<RootProvider
  search={{
    options: {
      api: '/base/api/search',
    },
  }}
>
  {children}
</RootProvider>
```

**Reply by nktnet1** [reply]
Ahh, I got it now. Thank you so much @fuma-nama!

I really appreciate all the pointers - have been able to implement everything I wanted.

🙏 🎉
